src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs

```diff
@@ -34,9 +32,7 @@ protected override async Task<ImmutableArray<ISymbol>> GetPreselectedSymbolsAsyn
             var recommender = context.GetLanguageService<IRecommendationService>();
             var typeInferrer = context.GetLanguageService<ITypeInferenceService>();
 
-            var inferredTypes = typeInferrer.InferTypes(context.SemanticModel, position, cancellationToken)
-                .Where(t => t.SpecialType != SpecialType.System_Void)
-                .ToSet();
+            var inferredTypes = context.InferredTypes.Where(t => t.SpecialType != SpecialType.System_Void).ToSet();
```

Comment:
Is it possible to use immutable set here?

Reply:
Another thought (not required):
Is it possible to change the InferredType in SyntaxContext to be lazy? So that we won't pay the cost if the Provider doesn't even use the inferred type.